### PR TITLE
Stable ABI

### PR DESCRIPTION
**Pull request by dcwatson**
* Use the stable Python ABI
* Return ~`memoryview`~ `bytearray` from compression methods
* ~Error when specifying incorrect size on decompression~
* Eliminate some boilerplate in the C module

### CODE REVIEW

**Comment by dcwatson** [comment]
I don't think this is going to work (yet). The biggest issue is that `Py_buffer` was not stabilized until 3.11. I could change the API to just accept `bytes`, or just read-only bytes-like objects (via the `y#` format), but would rather not.

**Comment by henryiii** [comment]
It was not added to the Stable ABI, that doesn't mean it wasn't stable.

File: src/deflate.c [hunk 1]
```diff
@@ -2,6 +2,9 @@
 #include "libdeflate.h"
 #include <Python.h>
 
+// Use the stable ABI
+#define Py_LIMITED_API 0x030b00f0
```
Comment (henryiii) [inline src/deflate.c]: This should be protected on Python version. You only want to define it for 3.11+. It also needs to be set before you include Python. You can communicate this via CMake.

Comment (henryiii) [reply src/deflate.c]: (Actually, the easiest way is to do the define from CMake)

Comment (dcwatson) [reply src/deflate.c]: But how will I know the Python version before including `<Python.h>`? Defining it after the include didn't actually remove the symbols.

Comment (dcwatson) [reply src/deflate.c]: And even if I add a CMake option, and conditionally set it during the build process based on which Python version its targeting, I'm still left with `_PyBytes_Resize`, meaning I'd also need to change the API to use `bytearray`. It's almost trivial to do so now, but I figured I'd get 0.7.0 out the door with the new build process.

Comment (henryiii) [reply src/deflate.c]: Getting 0.7 out is perfectly fine! I'm trying this out in a PR.

**Comment by henryiii** [comment]
> The biggest issue is that Py_buffer was not stabilized until 3.11.

It's the "Limited API" and "Stable ABI". Something can be a stable part of the API but not be part of the "Limited API".

**Pull request merged**